ndscan/fitting/fitting.py

```diff
@@ -0,0 +1,312 @@
+
+        lower_bounds = [self._param_bounds[param][0] for param in self._free_params]
+        upper_bounds = [self._param_bounds[param][1] for param in self._free_params]
+
+        p0 = [initial_values[param] for param in self._free_params]
+        p0 = [max(estimate, lower) for estimate, lower in zip(p0, lower_bounds)]
+        p0 = [min(estimate, upper) for estimate, upper in zip(p0, upper_bounds)]
+
+        scale_factors = [
+            self._scale_factors.get(param, 1) for param in self._free_params
+        ]
+        scale_factors = np.asarray(scale_factors)
+
+        p0 = np.asarray(p0) / scale_factors
+        lower_bounds = np.asarray(lower_bounds) / scale_factors
+        upper_bounds = np.asarray(upper_bounds) / scale_factors
+
+        p, p_cov = optimize.curve_fit(
+            f=functools.partial(self._func_free, scale_factors),
+            xdata=x,
+            ydata=y,
+            p0=p0,
+            sigma=y_err,
+            absolute_sigma=y_err is not None,
+            bounds=(lower_bounds, upper_bounds),
+        )
+
+        p *= scale_factors
+        p_cov *= scale_factors**2
+
+        self._p_cov = p_cov
+        p_err = np.sqrt(np.diag(p_cov))
+
+        self._p = {param: value for param, value in zip(self._free_params, p)}
+        self._p_err = {param: value for param, value in zip(self._free_params, p_err)}
+        self._p.update(self._fixed_params)
+        self._p_err.update({param: 0 for param in self._fixed_params.keys()})
+
+        self._calculate_derived_params()
+        assert set(self.get_params() + self.get_derived_params()) == set(self._p.keys())
+        return self._p, self._p_err
+
+    def _func_free(self, scale_factors, x, *args):
+        """Evaluate the fit function with given values for the free parameters."""
+        params = dict(self._fixed_params)
+
+        p = np.asarray(args)
+        p *= scale_factors
+        params.update({k: v for k, v in zip(self._free_params, p)})
+
+        return self.func(x, params)
+
+    def evaluate(self, x_fit: Union[np.array, int] = 100):
+        """Evaluates the function along x-fit and returns the tuple (x_fit, y_fit)
+        with the results.
+
+        `x_fit` may either be a scalar or an array. If it is a scalar it gives the
+        number of equally spaced points between `min(self.x)` `max(self.x)`. Otherwise
+        it gives the x-axis to use.
+        """
+        if np.isscalar(x_fit):
+            x_fit = np.linspace(np.min(self._x), np.max(self._x), x_fit)
+        y_fit = self.func(x_fit, self._p)
+        return x_fit, y_fit
+
+    @property
+    def residuals(self):
+        """Returns the fit residuals."""
+        return self._y - self._func_free(self._x, self._p)
+
+    @property
+    def chi2(self) -> Tuple[float, float]:
+        """ Chi-Squared test.
+
+        Assumes the values of `y` are independent and normally distributed.
+        Returns the tuple (Chi-Squared, p-value)
```

Comment:
I haven't had a closer look at this at all, but it might be valuable to offer some guidance here to users regarding the fact that our data is pretty much never actually normally distributed when working with ion readout data.

Reply:
mmm...I'm not sure that I have much intelligent to say here (but that probably mainly speaks to my limited knowledge of statistics) beyond "if the errors are well-approximated by normal then the Chi2 p-value is probably a decent estimator for the fit significance."

A couple of other comments here:
1. I was thinking of perhaps adding an additional `get_significance` (returns a number between 0 and 1) method to `FitBase` which would default to returning the Chi Squared p-value. That leaves the option open for people to do something cleverer if desired
2. I was also thinking of splitting `FitBase.fit` into two parts: a fit initialisation part and a part that does the fitting. The latter would default to least-squares fitting but allows people to do some other form of parameter estimation if desired. 
3. I was thinking of also letting `ExportedResult` take a result channel and a fit significance parameter. If the fit significance is above the threshold then the linked dataset is updated. This could also be done through a post-fit call back inside the `Analysis` class